Update guide on conda support (#7965)
readthedocs/readthedocs.org: docs/guides/conda.rst

This work was funded by `Clinical Graphics`_ -- many thanks for their support of Open Source.

.. _Clinical Graphics: https://www.clinicalgraphics.com/

Activating Conda
----------------

Conda support is available using a :doc:`../config-file/index`, see :ref:`config-file/v2:conda`.

Our :ref:`Docker images <builds:Docker images>` use Miniconda, a minimal conda installer.

> After specifying your project requirements using conda ``environment.yml`` files,

Reviewer: we allow setting only one file for now

```suggestion
After specifying your project requirements using a conda ``environment.yml`` file,
```

Read the Docs will create the environment (using ``conda env create``)
and add the core dependencies needed to build the documentation.

Creating the ``environment.yml``
--------------------------------

There are several ways of `exporting a conda environment`_:

- ``conda env export`` will produce a complete list of all the packages installed in the environment
  with their exact versions. This is the best option to ensure reproducibility,

>   but can create problems if done from a different operative system than the target machine.

Reviewer 2: Should we note the OS of our build servers? Perhaps point to the `builds.rst` page, and document it there, so we don't have to update it in multiple places?

Author: Good point, will add a reference there

- ``conda env export --from-history`` will only include packages that were explicitly requested,
  excluding the transitive dependencies. This is the best option to maximize cross-platform compatibility.
- And finally, you can also write it by hand. This allows for more fine-tuned control and overcomes

>   some limitations in the conda exporting capabilities.

Reviewer 3: I think this is _our_ preferred way and I suppose it's the one we want to recommend somehow. Why? Because otherwise, people install _all the requirements_ to _use the project_, instead only the requirements to _build the documentation_ --which could be minimal in comparison.

I know that there are lot of people using notebooks on their own documentation showing how their own package works. Those case are special and may want to treat them differently.

I'm not sure if this work is in the scope of this PR, tho. Probably is not, but worth to mention it.

Reviewer: Probably just adding a note like:  "Create an environment file only with the packages needed to build your docs for faster builds"

Author: And/or also, in the second method: "notice that this might include packages that are not needed to build your docs". Thanks for the suggestions!

Author: Done, have a look

For example, using the second method for an existing environment:

.. code-block::

    $ conda activate rtd38

>     (rtd38) $ conda env export --from-history | tee environment.yml

Reviewer: is there any problem with using `>`?

```suggestion
    (rtd38) $ conda env export --from-history > environment.yml
```

Author: I wanted to both show it on screen and save it to `environment.yml`, hence the use of `tee`

    name: rtd38
    channels:
      - defaults
      - conda-forge
    dependencies:
      - rasterio==1.2
      - python=3.8
      - pytorch-cpu=1.7

>     prefix: /home/juanlu/.mambaforge/envs/rtd38

Reviewer 3: Is this path showing `.mambaforge` because you are using mamba? If so, it may be better to change it to the conda default's dir to avoid confusions.

Author: Correct. Yes, I can manually change it to avoid confusion.

Author: Done!

> Read the Docs will override the ``name`` and ``prefix`` of the environment when creating it,

Reviewer: I know we override name here https://github.com/readthedocs/readthedocs.org/blob/29c23600d7ef2f2b63d90c900f0d13fff44eaec6/readthedocs/doc_builder/python_environments.py#L572-L573 but I don't see where we override prefix, or is it done implicitly?

Author: I think the prefix is overridden implicitly when changing the name, but to be 100 % sure I'd have to check the conda code.

so they can have any value, or not be present at all.

.. tip:: Bear in mind that ``rasterio==1.2`` will install version ``1.2.0``,

>    whereas ``python=3.8`` will fetch the latest ``3.8.*``,

Reviewer 3: It may be good to add something like "note the double `==`" --if that's what makes the difference here :smile:

Author: Good catch! Done

>    whereas ``python=3.8`` will fetch the latest ``3.8.*``,

Reviewer: ```suggestion
   whereas ``python=3.8`` will fetch the latest ``3.8.*`` version,
```

.. tip:: Bear in mind that ``rasterio==1.2`` (double ``==``) will install version ``1.2.0``,
   whereas ``python=3.8`` (single ``=``) will fetch the latest ``3.8.*`` version,

>    which is ``3.8.8`` at the time of writing.

Reviewer 2: 💯

   is not yet supported by default when using conda (see `this GitHub issue for discussion`_).
   If your project needs it, request that we enable the ``CONDA_APPEND_CORE_REQUIREMENTS``

>    :ref:`feature flag <guides/feature-flags:Feature Flags>`.

Reviewer 3: I hope we find a better solution soon. This is bad UX :disappointed: --maybe it worth to enable this feature by default on new projects and start testing it more? Besides fixing this issue it also speeds up the creation/installation of packages.

Same here; I think it's out of the scope of this PR, but I'm happy that we have a note in this document mentioning as a workaround.

Reviewer: +1 on testing it more to have it enable by default

Reviewer 2: Agreed. We should make it the default -- perhaps `future_default_true` for now? Then see if we can backport it?

   If your project needs it, request that we enable the ``CONDA_APPEND_CORE_REQUIREMENTS``
   :ref:`feature flag <guides/feature-flags:Feature Flags>`.

.. _this GitHub issue for discussion: https://github.com/readthedocs/readthedocs.org/issues/3829
.. _exporting a conda environment: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#sharing-an-environment

Effective use of channels
-------------------------

Conda packages are usually hosted on https://anaconda.org/, a registration-free artifact archive
maintained by Anaconda Inc. Contrary to what happens with the Python Package Index,
different users can potentially host the same package in the same repository,

> each of them using their own **channel**. Therefore, when installing a conda package,

Reviewer 2: ```suggestion
each of them using their own *channel*. Therefore, when installing a conda package,
```
```suggestion

Probably doesn't need bold emphasis. 
each of them using their own **channel**. Therefore, when installing a conda package,
```

> conda also needs to know which channels should use, and which ones take precedence.

Reviewer 2: ```suggestion
conda also needs to know which channels to use, and which ones take precedence.
```
```suggestion
conda also needs to know which channels should use, and which ones take precedence.
```

and subject to `Anaconda Terms of Service`_. It contains well-tested versions of the most widely used
packages. However, some packages are not available on the ``defaults`` channel,
and even if they are, they might not be on their lastest versions.

As an alternative, there are channels maintained by the community that have a broader selection
of packages and more up-to-date versions of them, the most popular one being ``conda-forge``.

To use the ``conda-forge`` channel when specifying your project dependencies, include it in the list
of ``channels`` in ``environment.yml``, and conda will rank them in order of appearance.
To maximize compatibility, we recommend putting ``conda-forge`` above ``defaults``.

.. tip:: If you want to opt out the ``defaults`` channel completely, replace it by ``nodefaults``

>    in the list of channels. See `the relevant conda docs`_ for more information.

Reviewer 3: Nice tip! I wasn't aware of this one.

> To maximize compatibility, we recommend putting ``conda-forge`` above ``defaults``.

Reviewer 2: Could use a code snippet showing this here.

Author: Done!

.. _Anaconda Terms of Service: https://www.anaconda.com/terms-of-service
.. _the relevant conda docs: https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html?highlight=nodefaults#creating-an-environment-file-manually

Mixing conda and pip packages
-----------------------------

There are valid reasons to use pip inside a conda environment: some dependency
might not be avaliable yet as a conda package in any channel,
or you might want to avoid precompiled binaries entirely.
In either case, it is possible to specify the subset of packages

> that will be installed with pip in the `environment.yml` file. For example:

Reviewer 3: ```suggestion
that will be installed with pip in the ``environment.yml`` file. For example:
```

.. code-block:: yaml

    name: rtd38
    channels:
      - defaults
      - conda-forge
    dependencies:
      - rasterio==1.2
      - python=3.8
      - pytorch-cpu=1.7
      - pip>=20.1  # pip is needed as dependency
      - pip:
        - black==20.8b1

>     prefix: /home/juanlu/.mambaforge/envs/rtd38

Reviewer: think we could remove the prefix key from the examples

Author: Good idea, done!
